icicle/include/icicle/fields/host_math.h

```diff
@@ -197,6 +199,167 @@ namespace host_math {
+        r[i + j] = madc_cc(m, n[j], r[i + j], c);
+      }
+      // Propagate the carry to the remaining sublimbs
+      for (int carry_idx = s + i; carry_idx < 2 * s; carry_idx++) {
+        if (c == 0) { break; }
+        r[carry_idx] = add_cc(r[carry_idx], c, c);
+      }
+    }
+  }
+
+  /**
+   * @brief Perform  SOS reduction on a number in montgomery representation \p t in range [0, \p n ^2-1] limiting it to
+   * the range [0,2 \p n -1].
+   * @param t Number to be reduced. Must be in montgomery rep, and in range [0, \p n ^2-1].
+   * @param n Field modulus.
+   * @param n_tag Number such that \p n * \p n_tag modR = -1
+   * @param r Array in which to store the result in its upper half (Lower half is data that would be removed by
+   * dividing by R = shifting NLIMBS down).
+   * @tparam NLIMBS Number of 32bit limbs required to represent a number in the field defined by n. R is 2^(NLIMBS*32).
+   */
+  template <unsigned NLIMBS>
+  static HOST_INLINE void
+  sos_mont_reduction_64(const uint64_t* t, const uint64_t* n, const uint64_t* n_tag, uint64_t* r)
+  {
+    const unsigned s = NLIMBS / 2; // Divide by 2 because NLIMBS is 32bit and this function is 64bit
+
+    // Copy t to r as t is read-only
+    for (int i = 0; i < 2 * s; i++) {
+      r[i] = t[i];
+    }
+
+    for (int i = 0; i < s; i++) {
+      uint64_t c = 0;
+      uint64_t m = r[i] * n_tag[0];
+
+      for (int j = 0; j < s; j++) {
+        // r[i+j] = addc_cc(r[i+j], m * n[j], c);
+        r[i + j] = madc_cc_64(m, n[j], r[i + j], c);
+      }
+      // Propagate the carry to the remaining sublimbs
+      for (int carry_idx = s + i; carry_idx < 2 * s; carry_idx++) {
+        if (c == 0) { break; }
+        r[carry_idx] = add_cc(r[carry_idx], c, c);
+      }
+    }
+  }
+
+  /**
+   * @brief Perform  SOS reduction on a number in montgomery representation \p t in range [0, \p n ^2-1] limiting it to
+   * the range [0,2 \p n -1].
+   * @param t Number to be reduced. Must be in montgomery rep, and in range [0, \p n ^2-1].
+   * @param n Field modulus.
+   * @param n_tag Number such that \p n * \p n_tag modR = -1
+   * @param r Array in which to store the result in its upper half (Lower half is data that would be removed by
+   * dividing by R = shifting NLIMBS down).
+   * @tparam NLIMBS Number of 32bit limbs required to represent a number in the field defined by n. R is 2^(NLIMBS*32).
+   */
+  template <unsigned NLIMBS, bool USE_32 = false>
+  static HOST_INLINE void sos_mont_reduction(
+    const storage<2 * NLIMBS>& t, const storage<NLIMBS>& n, const storage<NLIMBS>& n_tag, storage<2 * NLIMBS>& r)
+  {
+    static_assert(NLIMBS % 2 == 0 || NLIMBS == 1, "Odd number of limbs (That is not 1) is not supported\n");
+    if constexpr (USE_32) {
+      sos_mont_reduction_32<NLIMBS>(t.limbs, n.limbs, n_tag.limbs, r.limbs);
+      return;
+    } else if constexpr (NLIMBS == 1) {
+      sos_mont_reduction_32<NLIMBS>(t.limbs, n.limbs, n_tag.limbs, r.limbs);
+      return;
```

Comment:
why split the statements?

Reply:
for future optimization

Reply:
I don't see a scenario where NLIMBS=1 but USE_32=OFF and you can add this if else case in the future when you need it.